frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -44,23 +46,46 @@ public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyEx
   }
 
   @Override
-  public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
+  public void removePlaceVisit(String tripId, String placeUuid) throws PlaceVisitNotFoundException {
     Map<String, PlaceVisitModel> placesMap = placesByTripIdByPlaceId.get(tripId);
     if (placesMap == null) {
       throw new PlaceVisitNotFoundException(
-          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+          "PlaceVisit with id: '" + placeUuid + "' not found for trip '" + tripId + "'");
     }
 
-    if (placesMap.remove(placeId) == null) {
+    if (placesMap.remove(placeUuid) == null) {
       throw new PlaceVisitNotFoundException(
-          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+          "PlaceVisit with id: '" + placeUuid + "' not found for trip '" + tripId + "'");
     }
   }
 
   @Override
-  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeUuid) {
     return Optional.ofNullable(placesByTripIdByPlaceId.get(tripId))
-        .map(placesMap -> placesMap.get(placeId));
+        .map(placesMap -> placesMap.get(placeUuid));
+  }
+
+  @Override
+  public PlaceVisitModel updateUserMark(String tripId, String placeUuid,
+      PlaceVisitModel.UserMark newStatus) throws PlaceVisitNotFoundException {
+    AtomicReference<PlaceVisitModel> updatedPlaceVisit = new AtomicReference();
+
+    placesByTripIdByPlaceId.computeIfPresent(tripId, (tripKey, placesMap) -> {
+      PlaceVisitModel placeVisit = placesMap.get(placeUuid);
```

Comment:
This should be yet another `computeIfPresent` call to make sure the map is updated atomically. You can then use the return value to set the outer result:

```java
placesByTripIdByPlaceId.computeIfPresent(tripId, (tripKey, places) -> {
  PlaceVisitModel placeVisit = places.computeIfPresent(placeUuid, (placeId, place) ->
      place.toBuilder().setUserMark(newStatus).build());
  updatedPlaceVisit.set(placeVisit);
  return places;
});
```